File: src/AnnotationsReader/BasicAnnotationsReader.ts
import json5 from "json5";
import ts from "typescript";
import { AnnotationsReader } from "../AnnotationsReader";
import { Annotations } from "../Type/AnnotatedType";
import { symbolAtNode } from "../Utils/symbolAtNode";

export class BasicAnnotationsReader implements AnnotationsReader {
    private static requiresDollar = new Set<string>(["id", "comment", "ref"]);
    private static textTags = new Set<string>([
        "title",
        "description",
        "id",

        "format",
        "pattern",
        "ref",

        // New since draft-07:
        "comment",
        "contentMediaType",
        "contentEncoding",

        // Custom tag for if-then-else support.
        "discriminator",
    ]);
    private static jsonTags = new Set<string>([
        "minimum",
        "exclusiveMinimum",

        "maximum",
        "exclusiveMaximum",

        "multipleOf",

        "minLength",
        "maxLength",

        "minProperties",
        "maxProperties",

        "minItems",
        "maxItems",
        "uniqueItems",

        "propertyNames",
        "contains",
        "const",
        "examples",

        "default",

        // New since draft-07:
        "if",
        "then",
        "else",
        "readOnly",
        "writeOnly",

        // New since draft 2019-09:
        "deprecated",
    ]);

    public constructor(private extraTags?: Set<string>) {}

    public getAnnotations(node: ts.Node): Annotations | undefined {
        const symbol = symbolAtNode(node);
        if (!symbol) {
            return undefined;
        }

        const jsDocTags: ts.JSDocTagInfo[] = symbol.getJsDocTags();
        if (!jsDocTags || !jsDocTags.length) {
            return undefined;
        }

        const annotations = jsDocTags.reduce((result: Annotations, jsDocTag) => {
            const value = this.parseJsDocTag(jsDocTag);
            if (value !== undefined) {
                if (BasicAnnotationsReader.requiresDollar.has(jsDocTag.name)) {
                    result["$" + jsDocTag.name] = value;
                } else {
                    result[jsDocTag.name] = value;
                }
            }
            return result;
        }, {});

        return Object.keys(annotations).length ? annotations : undefined;
    }

    private parseJsDocTag(jsDocTag: ts.JSDocTagInfo): any {
        const isTextTag = BasicAnnotationsReader.textTags.has(jsDocTag.name);
        // Non-text tags without explicit value (e.g. `@deprecated`) default to `true`.
        const defaultText = isTextTag ? "" : "true";
        const text = jsDocTag.text?.map((part) => part.text).join("") || defaultText;

        if (isTextTag) {
            return text;
        } else if (BasicAnnotationsReader.jsonTags.has(jsDocTag.name)) {
            return this.parseJson(text) ?? text;
        } else if (this.extraTags?.has(jsDocTag.name)) {
            return this.parseJson(text) ?? text;
        } else {
            // Unknown jsDoc tag.
            return undefined;
        }
    }

    private parseJson(value: string): any {
        try {
            return json5.parse(value);
        } catch (e) {
            return undefined;
        }
    }
}

File: internal/parser/basic_annotations_reader.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// BasicAnnotationsReader reads schema keywords from JSDoc tags
// (src/AnnotationsReader/BasicAnnotationsReader.ts).
type BasicAnnotationsReader struct {
	extraTags map[string]bool
}

// requiresDollar tags are emitted with a "$" prefix ($id, $comment, $ref).
var requiresDollar = map[string]bool{
	"id":      true,
	"comment": true,
	"ref":     true,
}

// textTags carry their raw text value.
var textTags = map[string]bool{
	"title":       true,
	"description": true,
	"id":          true,

	"format":  true,
	"pattern": true,
	"ref":     true,

	// New since draft-07:
	"comment":          true,
	"contentMediaType": true,
	"contentEncoding":  true,

	// Custom tag for if-then-else support.
	"discriminator": true,
}

// jsonTags carry a JSON5-encoded value.
var jsonTags = map[string]bool{
	"minimum":          true,
	"exclusiveMinimum": true,

	"maximum":          true,
	"exclusiveMaximum": true,

	"multipleOf": true,

	"minLength": true,
	"maxLength": true,

	"minProperties": true,
	"maxProperties": true,

	"minItems":    true,
	"maxItems":    true,
	"uniqueItems": true,

	"propertyNames": true,
	"contains":      true,
	"const":         true,
	"examples":      true,

	"default": true,

	"required": true,

	// New since draft-07:
	"if":        true,
	"then":      true,
	"else":      true,
	"readOnly":  true,
	"writeOnly": true,

	// New since draft 2019-09:
	"deprecated": true,
}

// NewBasicAnnotationsReader creates a reader; extraTags is a set of
// additional tag names parsed like jsonTags.
func NewBasicAnnotationsReader(extraTags map[string]bool) *BasicAnnotationsReader {
	return &BasicAnnotationsReader{extraTags: extraTags}
}

func (r *BasicAnnotationsReader) GetAnnotations(node *ast.Node) types.Annotations {
	jsDocTags := symbolJSDocTags(node)
	if len(jsDocTags) == 0 {
		return nil
	}

	annotations := types.Annotations{}
	for _, jsDocTag := range jsDocTags {
		value, ok := r.parseJSDocTag(jsDocTag)
		if !ok {
			continue
		}
		if requiresDollar[jsDocTag.name] {
			annotations["$"+jsDocTag.name] = value
		} else {
			annotations[jsDocTag.name] = value
		}
	}

	if len(annotations) == 0 {
		return nil
	}
	return annotations
}

func (r *BasicAnnotationsReader) parseJSDocTag(jsDocTag jsDocTagInfo) (any, bool) {
	isTextTag := textTags[jsDocTag.name]
	text := jsDocTag.text
	if text == "" && !isTextTag {
		// Non-text tags without explicit value (e.g. `@deprecated`) default to `true`.
		text = "true"
	}

	if isTextTag {
		return text, true
	}
	parsed, err := ParseJSON5(text)
	if err != nil {
		parsed = text
	}
	if jsonTags[jsDocTag.name] || r.extraTags[jsDocTag.name] {
		return parsed, true
	}
	// Unknown jsDoc tag.
	return nil, false
}
